File: assets/backing-services/typescript/event-store-memory.ts
import type {
  AppendResult,
  CommittedEvent,
  Condition,
  ConditionalAppendResult,
  DomainEvent,
  EventStore,
  TaggedRead,
  TagQuery,
  TagsOf,
} from '../../application/ports/events.js';
import {
  appended,
  conditionConflict,
  defaultTagsOf,
  matchesQuery,
  NO_STREAM,
  recorded,
  versionConflict,
} from '../../application/ports/events.js';

/**
 * In-memory event-store adapter — for a quickstart with no infrastructure, and for tests that do not need
 * to prove concurrency. `make verify` runs the shared contract suite against this one, which is why the
 * repository gate needs no Docker.
 *
 * It satisfies the port's contract, but understand what it cannot do: being single-threaded, it cannot
 * genuinely race, so it CANNOT prove that two concurrent appends at the same version produce exactly one
 * winner — nor that two conditional appends against one boundary do. Those guarantees live in a real
 * store's unique constraint and isolation level, and `make test-integration` is where they are proved.
 * Ship on Postgres, demo on memory.
 *
 * Data is lost on restart, which for an event-sourced system means the entire truth is lost. Never
 * production.
 */

/**
 * What a connection is, when there is no database.
 *
 * It exists so the in-memory adapters can do the one thing the real ones do that matters most to the read
 * side: put an append, a view write and a checkpoint in **one transaction**. Build the event store and the
 * checkpoint store from the same instance — which is what `createInMemoryCheckpointStore(store)` does —
 * and `inUnitOfWork` covers all of it.
 *
 * Rollback is a restore from a shallow copy taken on entry. That is honest for what this is for: it proves
 * the *semantics* a test needs (a failed batch changes nothing) without pretending to be a transaction
 * manager. Events are frozen, so copying the list is enough.
 */
export type InMemoryDatabase = {
  log: CommittedEvent[];
  /**
   * Global position to the tags it was indexed by — the derived index, kept beside the log rather than on
   * the event, exactly as the SQL adapters keep it in a table of its own.
   */
  tags: Map<number, readonly string[]>;
  checkpoints: Map<string, number>;
  leases: Map<string, { owner: string; expiresAt: Date }>;
  head(): number;
  inUnitOfWork<T>(work: () => Promise<T>): Promise<T>;
};

export function createInMemoryDatabase(): InMemoryDatabase {
  const database: InMemoryDatabase = {
    log: [],
    tags: new Map(),
    checkpoints: new Map(),
    leases: new Map(),
    head: () => database.log.at(-1)?.globalPosition ?? 0,
    // Nesting needs no counter: the mutation happens in place, so a rollback is a restore, and an inner
    // block restoring its own snapshot leaves the outer one's intact.
    inUnitOfWork: async <T>(work: () => Promise<T>): Promise<T> => {
      const log = database.log.slice();
      const tags = new Map(database.tags);
      const checkpoints = new Map(database.checkpoints);
      const leases = new Map(database.leases);
      try {
        return await work();
      } catch (error) {
        database.log = log;
        database.tags = tags;
        database.checkpoints = checkpoints;
        database.leases = leases;
        throw error;
      }
    },
  };
  return database;
}

export type InMemoryEventStore = EventStore & { readonly database: InMemoryDatabase };

export function createInMemoryEventStore(
  tagsOf: TagsOf = defaultTagsOf,
  database: InMemoryDatabase = createInMemoryDatabase(),
): InMemoryEventStore {
  let tagging = tagsOf;

  const streamEvents = (streamId: string): CommittedEvent[] =>
    database.log.filter((event) => event.streamId === streamId);

  const versionOf = (streamId: string): number =>
    streamEvents(streamId).at(-1)?.version ?? NO_STREAM;

  const record = (event: DomainEvent, streamId: string, version: number): void => {
    const globalPosition = database.head() + 1;
    database.log.push({
      ...event,
      streamId,
      version,
      globalPosition,
      recordedAt: new Date().toISOString(),
    });
    // No entry at all when the tagging function returns nothing, rather than an empty one: the SQL
    // adapters give such an event no rows, and "indexed" has to mean the same thing in every adapter
    // or reindexTags answers differently depending on the store.
    const tags = tagging(event);
    if (tags.length > 0) database.tags.set(globalPosition, [...tags]);
  };

  const readTagged = async (query: TagQuery, after = 0, until = 0): Promise<TaggedRead> => {
    const ceiling = until === 0 ? database.head() : until;
    return {
      events: database.log.filter(
        (event) =>
          event.globalPosition > after &&
          event.globalPosition <= ceiling &&
          matchesQuery(query, event, database.tags.get(event.globalPosition) ?? []),
      ),
      head: ceiling,
    };
  };

  const reindexTags = async (fromPosition = 0): Promise<number> =>
    database.inUnitOfWork(async () => {
      let indexed = 0;
      for (const event of database.log) {
        if (event.globalPosition < fromPosition) continue;
        if (database.tags.has(event.globalPosition)) continue;
        const tags = tagging(event);
        if (tags.length === 0) continue;
        database.tags.set(event.globalPosition, [...tags]);
        indexed += 1;
      }
      return indexed;
    });

  return {
    database,

    read: async (streamId: string): Promise<readonly CommittedEvent[]> =>
      streamEvents(streamId).slice(),

    append: async (
      streamId: string,
      expectedVersion: number,
      events: readonly DomainEvent[],
    ): Promise<AppendResult> => {
      const actualVersion = versionOf(streamId);
      if (actualVersion !== expectedVersion) return versionConflict(actualVersion);

      let version = actualVersion;
      await database.inUnitOfWork(async () => {
        for (const event of events) {
          version += 1;
          record(event, streamId, version);
        }
      });
      return appended(version);
    },

    readAll: async function* (fromPosition: number): AsyncIterable<CommittedEvent> {
      for (const event of database.log.filter((entry) => entry.globalPosition >= fromPosition))
        yield event;
    },

    inUnitOfWork: database.inUnitOfWork,

    head: async () => database.head(),
    readTagged,

    appendIf: async (
      condition: Condition,
      events: readonly DomainEvent[],
    ): Promise<ConditionalAppendResult> => {
      const matching = await readTagged(condition.query, condition.after);
      if (matching.events.length > 0) return conditionConflict(database.head());

      await database.inUnitOfWork(async () => {
        for (const event of events) {
          // Each event still lands at the next version of the stream it names, so a conditionally
          // appended event is readable by everything written against `read` and `append`. The boundary
          // changed; the log did not.
          record(event, event.streamId, versionOf(event.streamId) + 1);
        }
      });
      return recorded(database.head());
    },

    reindexTags,

    retag: async (next: TagsOf): Promise<number> =>
      database.inUnitOfWork(async () => {
        tagging = next;
        database.tags.clear();
        return reindexTags();
      }),
  };
}

File: assets/backing-services/python/event_store_memory.py
"""In-memory event-store adapter.

For a quickstart with no infrastructure, and for tests that do not need to prove concurrency.
`make verify` runs the shared contract suite against this one, which is why the repository gate
needs no Docker.

It satisfies the port's contract, but understand what it cannot do: being single-threaded, it
cannot genuinely race, so it CANNOT prove that two concurrent appends at the same version
produce exactly one winner — nor that two conditional appends against one boundary do. Those
guarantees live in a real store's unique constraint and isolation level, and
`make test-integration` is where they are proved. Ship on Postgres, demo on memory.

Data is lost when the process ends, which for an event-sourced system means the entire truth
is lost. Never production.
"""

from __future__ import annotations

from collections.abc import Iterator, Sequence
from contextlib import AbstractContextManager, contextmanager
from datetime import UTC, datetime

from delivery_starter.application.ports.events import (
    NO_STREAM,
    Appended,
    AppendResult,
    CommittedEvent,
    Condition,
    ConditionalAppendResult,
    ConditionConflict,
    DomainEvent,
    EventStore,
    Recorded,
    TaggedRead,
    TagQuery,
    TagsOf,
    VersionConflict,
    default_tags_of,
)


class InMemoryDatabase:
    """What a connection is, when there is no database.

    It exists so the in-memory adapters can do the one thing the real ones do that matters most
    to the read side: put an append, a view write and a checkpoint in **one transaction**. Build
    the event store and the checkpoint store from the same instance — which is what
    `create_in_memory_checkpoint_store(store)` does — and `unit_of_work` covers all of it.

    Rollback is a restore from a shallow copy taken on entry. That is honest for what this is
    for: it proves the *semantics* a test needs (a failed batch changes nothing) without
    pretending to be a transaction manager. Events are frozen, so copying the list is enough;
    nothing here is shared with anything that mutates it afterwards.
    """

    def __init__(self) -> None:
        self.log: list[CommittedEvent] = []
        #: Global position to the tags it was indexed by — the derived index, kept beside the log
        #: rather than on the event, exactly as the SQL adapters keep it in a table of its own.
        self.tags: dict[int, tuple[str, ...]] = {}
        self.checkpoints: dict[str, int] = {}
        self.leases: dict[str, tuple[str, datetime]] = {}
        self._depth = 0

    @contextmanager
    def unit_of_work(self) -> Iterator[None]:
        """Re-entrant: the outermost block is the one that can roll back."""
        if self._depth:
            self._depth += 1
            try:
                yield
            finally:
                self._depth -= 1
            return
        snapshot = (list(self.log), dict(self.tags), dict(self.checkpoints), dict(self.leases))
        self._depth = 1
        try:
            yield
        except BaseException:
            self.log, self.tags, self.checkpoints, self.leases = (
                snapshot[0],
                snapshot[1],
                snapshot[2],
                snapshot[3],
            )
            raise
        finally:
            self._depth = 0

    @property
    def head(self) -> int:
        return self.log[-1].global_position if self.log else 0


class InMemoryEventStore(EventStore):
    def __init__(
        self,
        database: InMemoryDatabase | None = None,
        tags_of: TagsOf = default_tags_of,
    ) -> None:
        #: Public, because the checkpoint-store adapter is built from it: two in-memory adapters
        #: sharing one of these is the equivalent of two SQL adapters sharing one connection.
        self.database = database if database is not None else InMemoryDatabase()
        self._tags_of = tags_of

    def _stream(self, stream_id: str) -> list[CommittedEvent]:
        return [event for event in self.database.log if event.stream_id == stream_id]

    def _next_position(self) -> int:
        return self.database.head + 1

    def _record(self, event: DomainEvent, stream_id: str, version: int) -> CommittedEvent:
        committed = CommittedEvent(
            type=event.type,
            schema_version=event.schema_version,
            stream_id=stream_id,
            payload=dict(event.payload),
            occurred_at=event.occurred_at,
            actor=event.actor,
            correlation_id=event.correlation_id,
            causation_id=event.causation_id,
            version=version,
            global_position=self._next_position(),
            recorded_at=datetime.now(UTC).isoformat(),
        )
        self.database.log.append(committed)
        tags = tuple(self._tags_of(event))
        if tags:
            # No entry at all when the tagging function returns nothing, rather than an empty one:
            # the SQL adapters give such an event no rows, and "indexed" has to mean the same thing
            # in every adapter or `reindex_tags` answers differently depending on the store.
            self.database.tags[committed.global_position] = tags
        return committed

    def read(self, stream_id: str) -> tuple[CommittedEvent, ...]:
        return tuple(self._stream(stream_id))

    def append(
        self,
        stream_id: str,
        expected_version: int,
        events: Sequence[DomainEvent],
    ) -> AppendResult:
        stream = self._stream(stream_id)
        actual_version = stream[-1].version if stream else NO_STREAM
        if actual_version != expected_version:
            return VersionConflict(actual_version)

        version = actual_version
        with self.database.unit_of_work():
            for event in events:
                version += 1
                self._record(event, stream_id, version)
        return Appended(version)

    def read_all(self, from_position: int) -> Iterator[CommittedEvent]:
        for event in self.database.log:
            if event.global_position >= from_position:
                yield event

    def unit_of_work(self) -> AbstractContextManager[None]:
        return self.database.unit_of_work()

    def head(self) -> int:
        return self.database.head

    def read_tagged(self, query: TagQuery, after: int = 0, until: int = 0) -> TaggedRead:
        ceiling = until or self.database.head
        return TaggedRead(
            events=tuple(
                event
                for event in self.database.log
                if after < event.global_position <= ceiling
                and query.matches(event, self.database.tags.get(event.global_position, ()))
            ),
            head=ceiling,
        )

    def append_if(
        self,
        condition: Condition,
        events: Sequence[DomainEvent],
    ) -> ConditionalAppendResult:
        if self.read_tagged(condition.query, condition.after).events:
            return ConditionConflict(self.database.head)
        if not events:
            return Recorded(self.database.head)
        with self.database.unit_of_work():
            for event in events:
                # Each event still lands at the next version of the stream it names, so a
                # conditionally appended event is readable by everything written against
                # `read` and `append`. The boundary changed; the log did not.
                stream = self._stream(event.stream_id)
                self._record(
                    event,
                    event.stream_id,
                    (stream[-1].version if stream else NO_STREAM) + 1,
                )
        return Recorded(self.database.head)

    def reindex_tags(self, from_position: int = 0) -> int:
        indexed = 0
        with self.database.unit_of_work():
            for event in self.database.log:
                if event.global_position < from_position:
                    continue
                if event.global_position in self.database.tags:
                    continue
                tags = tuple(self._tags_of(_as_domain(event)))
                if not tags:
                    continue
                self.database.tags[event.global_position] = tags
                indexed += 1
        return indexed

    def retag(self, tags_of: TagsOf) -> int:
        with self.database.unit_of_work():
            self._tags_of = tags_of
            self.database.tags.clear()
            return self.reindex_tags()


def _as_domain(event: CommittedEvent) -> DomainEvent:
    """A stored event as the tagging function sees it.

    `tags_of` is written against `DomainEvent`, so it can be called on an event before it is
    appended *and* on one read back out of the log during a reindex. The two must agree —
    otherwise an index built by a rebuild differs from the one built by appends, and nothing
    would notice.
    """
    return DomainEvent(
        type=event.type,
        schema_version=event.schema_version,
        stream_id=event.stream_id,
        payload=event.payload,
        occurred_at=event.occurred_at,
        actor=event.actor,
        correlation_id=event.correlation_id,
        causation_id=event.causation_id,
    )


def create_in_memory_event_store(
    tags_of: TagsOf = default_tags_of,
    database: InMemoryDatabase | None = None,
) -> InMemoryEventStore:
    """Both arguments are optional, and a call with neither is the quickstart it always was.

    Pass `tags_of` where this project's events are findable by more than their stream, and
    `database` to build a checkpoint store on the same "connection".
    """
    return InMemoryEventStore(database=database, tags_of=tags_of)
